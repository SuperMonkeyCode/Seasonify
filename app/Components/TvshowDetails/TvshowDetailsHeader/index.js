/* @flow */
import { connect } from 'react-redux';

import { editActions, editSelectors } from 'Store/editStore';
import { tvshowActions, tvshowSelectors } from 'Store/tvshowStore';
import { uiActions } from 'Store/uiStore';

import TvshowDetailsHeader from './TvshowDetailsHeader';

export default connect(
  (state, { tvshowId }) => ({
    isEditing: editSelectors.isEditing(state),
    editedObject: editSelectors.editedObject(state),
    tvshow: tvshowSelectors.getTvshow(state, { tvshowId }),
  }),
  {
    tvshowDelete: tvshowActions.tvshowDelete,
    tvshowUpdate: tvshowActions.tvshowUpdate,
    messageToast: uiActions.messageToast,
    editStart: editActions.editStart,
    editEnd: editActions.editEnd,
  },
)(TvshowDetailsHeader);
