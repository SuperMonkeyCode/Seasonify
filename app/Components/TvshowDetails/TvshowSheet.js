/* @flow */
import { Content, Form, Item, Input, Icon, List, ListItem, Text } from 'native-base';
import { pure } from 'recompose';
import React from 'react';

type Props = {
  /* parent */
  name: string,
  year: number,
  localizedName?: string,
  isEditing: boolean,
  edit: Object,
  onChangeName: Function,
  /* connect */
  /* HOC */
};

const enhance = pure;

function TvshowSheet({
  name,
  year,
  localizedName,
  isEditing,
  edit,
  onChangeName,
}: Props) {
  return (
    <Content>
      <Form>
        <Item regular>
          {isEditing && <Icon name="create" />}
          <Input
            disabled={!isEditing}
            placeHolder="Name"
            value={'name' in edit ? edit.name : name}
            onChangeText={onChangeName}
            autoCapitalize="words"
          />
        </Item>
      </Form>
      <List>
        {localizedName &&
          <ListItem>
            <Text note>
              {localizedName}
            </Text>
          </ListItem>}
        <ListItem>
          <Text note>
            {year}
          </Text>
        </ListItem>
      </List>
    </Content>
  );
}

export default enhance(TvshowSheet);
