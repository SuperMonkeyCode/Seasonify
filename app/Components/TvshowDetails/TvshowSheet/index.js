/* @flow */
export { default } from './TvshowSheet';
