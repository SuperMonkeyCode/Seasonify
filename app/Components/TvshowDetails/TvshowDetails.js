/* @flow */
import { Col, Container, Content, Grid } from 'native-base';
import { compose, pure, withHandlers } from 'recompose';
import React from 'react';

import { Metrics } from 'Themes';
import Loading from 'Components/Loading';
import Poster from 'Components/Poster';
import SingleFAB from 'Components/SingleFAB';

import SeasonList from './SeasonList';
import TvshowSheet from './TvshowSheet';

type Props = {
  /* parent */
  navigation: Object,
  /* connect */
  tvshow: Tvshow,
  isEditing: boolean,
  editedObject: Object,
  editUpdate: Function,
  seasonsRefresh: Function,
  isSpinning: boolean,
  /* HOC */
  handleChangeName: Function,
  handleFAB: Function,
};

const enhance = compose(
  pure,
  withHandlers({
    handleChangeName: ({ navigation, editUpdate }: Props) => (name: string) => {
      editUpdate({ id: navigation.state.params.tvshowId, name });
    },
    handleFAB: ({ tvshow, seasonsRefresh }: Props) => () => {
      seasonsRefresh(tvshow.id);
    },
  }),
);

function TvshowDetails({
  navigation,
  tvshow,
  isEditing,
  editedObject,
  isSpinning,
  handleChangeName,
  handleFAB,
}: Props) {
  if (tvshow) {
    return (
      <Container>
        <Content>
          <Grid>
            <Col size={Metrics.columnLeft}>
              {tvshow.poster && <Poster url={tvshow.poster} />}
            </Col>
            <Col size={Metrics.columnRight}>
              <TvshowSheet
                name={tvshow.name}
                localizedName={tvshow.localizedName}
                year={tvshow.year}
                edit={editedObject}
                isEditing={isEditing}
                onChangeName={handleChangeName}
              />
            </Col>
          </Grid>
          <SeasonList
            tvshowId={tvshow.id}
            seasons={tvshow.seasons}
            navigation={navigation}
          />
        </Content>
        <SingleFAB icon="refresh" onPress={handleFAB} />
        <Loading position="nextToFAB" isSpinning={isSpinning} />
      </Container>
    );
  }
  return <Container />;
}

export default enhance(TvshowDetails);
