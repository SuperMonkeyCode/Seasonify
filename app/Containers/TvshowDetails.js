/* @flow */
import { Col, Container, Content, Grid, View } from 'native-base';
import { connect } from 'react-redux';
import React from 'react';

import type { Tvshow } from 'Types';
import { Metrics } from 'Themes';
import Poster from 'Components/Poster';
import SeasonList from 'Components/SeasonList';
import SingleFAB from 'Components/SingleFAB';
import TvshowSheet from 'Components/TvshowSheet';
import editState from 'State/editState';
import tv from 'State/tvshowState';

import Loading from './Loading';
import TvshowDetailsHeader from './TvshowDetailsHeader';

const mapStateToProps = (state, ownProps) => ({
  tvshow: tv.selectors.getTvshow(state, ownProps.navigation.state.params.tvshowId),
  isEditing: editState.selectors.isEditing(state),
  editedObject: editState.selectors.editedObject(state),
});

const mapActionsToProps = {
  editUpdate: editState.actions.editUpdate,
  seasonsRefresh: tv.actions.seasonsRefresh,
};

export function TvshowDetails({
  navigation,
  tvshow,
  isEditing,
  editedObject,
  editUpdate,
  seasonsRefresh,
}: {
  navigation: Object,
  tvshow: Tvshow,
  isEditing: boolean,
  editedObject: Object,
  editUpdate: Function,
  seasonsRefresh: Function,
}) {
  const { tvshowId } = navigation.state.params;

  const onChangeName = name => {
    editUpdate({ id: tvshowId, name });
  };

  const onFAB = () => {
    seasonsRefresh(tvshow.id);
  };

  if (tvshow) {
    return (
      <Container>
        <Content>
          <Grid>
            <Col size={Metrics.columnLeft}>
              <Poster url={tvshow.poster} />
            </Col>
            <Col size={Metrics.columnRight}>
              <TvshowSheet
                name={tvshow.name}
                localizedName={tvshow.localizedName}
                year={tvshow.year}
                edit={editedObject}
                isEditing={isEditing}
                onChangeName={onChangeName}
              />
            </Col>
          </Grid>
          <SeasonList seasons={tvshow.seasons} />
        </Content>
        <SingleFAB icon="refresh" onPress={onFAB} />
        <Loading color="lightblue" />
      </Container>
    );
  }
  return <Container />;
}

TvshowDetails.navigationOptions = ({ navigation }) => ({
  header: (
    <TvshowDetailsHeader
      tvshowId={navigation.state.params.tvshowId}
      navigate={navigation.navigate}
    />
  ),
});

export default connect(mapStateToProps, mapActionsToProps)(TvshowDetails);
