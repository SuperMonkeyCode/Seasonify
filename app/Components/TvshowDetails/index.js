/* @flow */
import { connect } from 'react-redux';
import React from 'react';

import { editActions, editSelectors } from 'Store/editStore';
import { tvshowActions, tvshowSelectors } from 'Store/tvshowStore';
import { uiSelectors } from 'Store/uiStore';
import TvshowDetailsHeader from 'Components/TvshowDetailsHeader';

import TvshowDetails from './TvshowDetails';

TvshowDetails.navigationOptions = ({ navigation }) => ({
  header: (
    <TvshowDetailsHeader
      tvshowId={navigation.state.params.tvshowId}
      navigation={navigation}
    />
  ),
});

export default connect(
  (state, { navigation: { state: { params: { tvshowId } } } }) => ({
    tvshow: tvshowSelectors.getTvshow(state, { tvshowId }),
    isEditing: editSelectors.isEditing(state),
    editedObject: editSelectors.editedObject(state),
    isSpinning: uiSelectors.isSpinning(state),
  }),
  {
    editUpdate: editActions.editUpdate,
    seasonsRefresh: tvshowActions.seasonsRefresh,
  },
)(TvshowDetails);
