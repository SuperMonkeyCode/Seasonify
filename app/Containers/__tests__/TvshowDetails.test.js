import React from 'react';
import { TvshowDetails } from '../TvshowDetails';

function setup(specificProps = {}) {
  const props = {
    navigation: {
      navigate: jest.fn(),
      state: { params: { tvshowId: 'abc123' } },
    },
    tvshow: {
      id: 'abc123',
      name: 'A tvshow',
      localizedName: 'Une s√©rie',
      poster: 'http://poster.png',
      year: 2017,
      seasons: {
        id: 'xxx999',
        allocine: 987,
        year: 2016,
        episodes: 1,
      },
    },
    isEditing: false,
    editedObject: {},
    editUpdate: jest.fn(),
    seasonsRefresh: jest.fn(),
    ...specificProps,
  };
  const component = shallow(<TvshowDetails {...props} />);
  return {
    component,
    props,
  };
}

describe('Rendering when default state', () => {
  const { component } = setup();
  it('should match', () => {
    expect(component).toMatchSnapshot();
  });
});

describe('Rendering when no tvshow', () => {
  const { component } = setup({ tvshow: undefined });
  it('should match', () => {
    expect(component).toMatchSnapshot();
  });
});

/* ========= Events & Functions ========= */

describe('Events & Functions', () => {
  const { component, props } = setup();
  const TvshowSheet = component.find('TvshowSheet');
  const SingleFAB = component.find('SingleFAB');

  describe('when calling onChangeName', () => {
    TvshowSheet.prop('onChangeName')('New name');

    it('calls editUpdate', () => {
      expect(props.editUpdate).toBeCalledWith({
        id: props.tvshow.id,
        name: 'New name',
      });
    });
  });

  describe('when calling onPress', () => {
    SingleFAB.prop('onPress')();

    it('calls seasonsRefresh', () => {
      expect(props.seasonsRefresh).toBeCalledWith(props.tvshow.id);
    });
  });
});
