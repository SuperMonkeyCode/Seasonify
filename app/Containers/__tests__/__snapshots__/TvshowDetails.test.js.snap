// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rendering when default state should match 1`] = `
<Styled(Container)>
  <Styled(Content)>
    <GridNB>
      <ColumnNB
        size={38}
      >
        <Poster
          url="http://poster.png"
        />
      </ColumnNB>
      <ColumnNB
        size={62}
      >
        <TvshowSheet
          edit={Object {}}
          isEditing={false}
          localizedName="Une sÃ©rie"
          name="A tvshow"
          onChangeName={[Function]}
          year={2017}
        />
      </ColumnNB>
    </GridNB>
    <SeasonList
      seasons={
        Object {
          "allocine": 987,
          "episodes": 1,
          "id": "xxx999",
          "year": 2016,
        }
      }
    />
  </Styled(Content)>
  <SingleFAB
    icon="refresh"
    onPress={[Function]}
  />
  <Connect(Loading)
    color="lightblue"
  />
</Styled(Container)>
`;

exports[`Rendering when no tvshow should match 1`] = `<Styled(Container) />`;
